hugr-core/src/extension.rs

```diff
@@ -372,14 +410,63 @@ impl Extension {
         }
     }
 
-    /// Extend the requirements of this extension with another set of extensions.
-    pub fn with_reqs(self, extension_reqs: impl Into<ExtensionSet>) -> Self {
-        Self {
-            extension_reqs: self.extension_reqs.union(extension_reqs.into()),
-            ..self
+    /// Creates a new extension wrapped in an [`Arc`].
+    ///
+    /// The closure lets us use a weak reference to the arc while the extension
+    /// is being built. This is necessary for calling [`Extension::add_op`] and
+    /// [`Extension::add_type`].
+    pub fn new_arc(
+        name: ExtensionId,
+        version: Version,
+        init: impl FnOnce(&mut Extension, &Weak<Extension>),
+    ) -> Arc<Self> {
+        Arc::new_cyclic(|extension_ref| {
+            let mut ext = Self::new(name, version);
+            init(&mut ext, extension_ref);
+            ext
+        })
+    }
+
+    /// Creates a new extension wrapped in an [`Arc`], using a fallible
+    /// initialization function.
+    ///
+    /// The closure lets us use a weak reference to the arc while the extension
+    /// is being built. This is necessary for calling [`Extension::add_op`] and
+    /// [`Extension::add_type`].
+    pub fn try_new_arc<E>(
+        name: ExtensionId,
+        version: Version,
+        init: impl FnOnce(&mut Extension, &Weak<Extension>) -> Result<(), E>,
+    ) -> Result<Arc<Self>, E> {
+        // Annoying hack around not having `Arc::try_new_cyclic` that can return
+        // a Result.
+        // https://github.com/rust-lang/rust/issues/75861#issuecomment-980455381
+        //
+        // When there is an error, we store it in `error` and return it at the
+        // end instead of the partially-initialized extension.
+        let mut error = None;
+        let ext = Arc::new_cyclic(|extension_ref| {
+            let mut ext = Self::new(name, version);
+            match init(&mut ext, extension_ref) {
+                Ok(_) => ext,
+                Err(e) => {
+                    error = Some(e);
+                    ext
+                }
+            }
+        });
+        match error {
+            Some(e) => Err(e),
+            None => Ok(ext),
         }
     }
 
+    /// Extend the requirements of this extension with another set of extensions.
+    pub fn set_reqs(&mut self, extension_reqs: impl Into<ExtensionSet>) {
```

Comment:
nit: "set" doesn't sound like extending the existing reqs to me. Maybe something like `add_reqs`?

Reply:
Renamed it to `add_requirements`